src/engine/Match_rules.ml

```diff
@@ -101,53 +92,109 @@ let check ~match_hook ~timeout ~timeout_threshold (xconf : Match_env.xconfig)
-                             match_hook xconf xtarget
-                       | `Taint _ as mode ->
-                           (* TODO: 'debug_taint' should just be part of 'res'
-                            * (i.e., add a "debugging" field to 'Report.match_result'). *)
-                           let res, _TODO_debug_taint =
-                             Match_tainting_mode.check_rule { r with mode }
-                               match_hook xconf xtarget
-                           in
-                           res
-                       | `Extract extract_spec ->
-                           Match_search_mode.check_rule
-                             { r with mode = `Search extract_spec.formula }
-                             match_hook xconf xtarget)
-                 in
-                 match match_result with
-                 | Some res -> Left res
-                 (* Note that because we now parse lazily a file, this rule timeout
-                  * can actually correspond to a parsing file timeout. *)
-                 | None ->
-                     incr cnt_timeout;
-                     if
-                       timeout_threshold > 0
-                       && !cnt_timeout >= timeout_threshold
-                     then raise File_timeout;
-                     let loc = Parse_info.first_loc_of_file file in
-                     let error =
-                       E.mk_error ~rule_id:(Some rule_id) loc "" Out.Timeout
-                     in
-                     Left
-                       (RP.make_match_result []
-                          (Report.ErrorSet.singleton error)
-                          (RP.empty_rule_profiling r))))
+           match r.R.mode with
+           | _ when not relevant_rule -> Right3 r
+           | `Taint _ as mode -> Left3 { r with mode }
+           | (`Extract _ | `Search _) as mode -> Middle3 { r with mode })
+  in
+  (* Taint rules are only relevant to each other if they are meant to be
+     analyzing the same language.
+     So we group the taint rules by common language, before passing them
+     to [Match_tainting_mode.check_rules].
+  *)
+  let relevant_taint_rules_groups =
+    relevant_taint_rules
+    |> Common.map (fun r -> (r.R.languages, r))
+    |> Common.group_assoc_bykey_eff |> Common.map snd
+  in
+  (relevant_taint_rules_groups, relevant_nontaint_rules, skipped_rules)
+
+(* Given a thunk [f] that computes the results of running the engine on a single
+    rule, this function simply instruments the computation on a single rule with
+    some boilerplate logic, like setting the last matched rule, timing out if
+    it takes too long, and producing a faulty match result in that case.
+
+    In particular, we need this to call [Match_tainting_mode.check_rules], which
+    will iterate over each rule in a different place, and so needs access to this
+    logic.
+*)
+let per_rule_boilerplate_fn ~timeout ~timeout_threshold file rule f =
+  let cnt_timeout = ref 0 in
```

Comment:
I guess you need to put `fun f -> ...` after this ?

Reply:
Otherwise each invocation of this `per_rule_boilerplate_fn` below will be creating a new `ref` initialized to `0`.
```ocaml
let per_rule_boilerplate_fn =
     per_rule_boilerplate_fn ~timeout ~timeout_threshold file
   in
```

Reply:
aw man, i used to explicitly teach people not to do this too...

Reply:
whoops